src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -0,0 +1,842 @@
+                    ValidateWritingProperty();
+                    UpdateBitStackOnStart(token);
+                }
+                idx = WritePropertyNameMinimized(ref propertyName);
+            }
+
+            if (1 > _buffer.Length - idx)
+            {
+                AdvanceAndGrow(idx, 1);
+                idx = 0;
+            }
+
+            _buffer[idx++] = token;
+
+            Advance(idx);
+        }
+
+        private void WriteStartEscapeProperty(ref ReadOnlySpan<char> propertyName, byte token, int firstEscapeIndexProp)
+        {
+            Debug.Assert(int.MaxValue / MaxExpansionFactorWhileEscaping >= propertyName.Length);
+
+            char[] propertyArray = null;
+
+            int length = firstEscapeIndexProp + MaxExpansionFactorWhileEscaping * (propertyName.Length - firstEscapeIndexProp);
+            Span<char> span;
+            if (length > StackallocThreshold)
+            {
+                propertyArray = ArrayPool<char>.Shared.Rent(length);
+                span = propertyArray;
+            }
+            else
+            {
+                // Cannot create a span directly since the span gets exposed outside this method.
+                unsafe
+                {
+                    char* ptr = stackalloc char[length];
+                    span = new Span<char>(ptr, length);
+                }
+            }
+            JsonWriterHelper.EscapeString(ref propertyName, ref span, firstEscapeIndexProp, out int written);
+            propertyName = span.Slice(0, written);
+
+            WriteStartByOptions(ref propertyName, token);
+
+            if (propertyArray != null)
+                ArrayPool<char>.Shared.Return(propertyArray);
+        }
+
+        /// <summary>
+        /// Writes the end of a JSON array.
+        /// </summary>
+        /// <exception cref="JsonWriterException">
+        /// Thrown if this would result in an invalid JSON to be written (while validation is enabled).
+        /// </exception>
+        public void WriteEndArray()
+        {
+            WriteEnd(JsonConstants.CloseBracket);
+            _tokenType = JsonTokenType.EndArray;
+        }
+
+        /// <summary>
+        /// Writes the end of a JSON object.
+        /// </summary>
+        /// <exception cref="JsonWriterException">
+        /// Thrown if this would result in an invalid JSON to be written (while validation is enabled).
+        /// </exception>
+        public void WriteEndObject()
+        {
+            WriteEnd(JsonConstants.CloseBrace);
+            _tokenType = JsonTokenType.EndObject;
+        }
+
+        private void WriteEnd(byte token)
+        {
+            _currentDepth |= 1 << 31;
+            _currentDepth--;
+
+            // Necessary if WriteEndX is called without a corresponding WriteStartX first.
+            // Checking for int.MaxValue because int.MinValue - 1 = int.MaxValue
+            if (_currentDepth == int.MaxValue)
+            {
+                _currentDepth = 0;
```

Comment:
If I write the wrong token and an exception is thrown, why does _currentDepth move?  Field writes shouldn't happen until after state validation.

Reply:
What happens if SkipValidation is true and writing a wrong token does not throw an exception? That is when _currentDepth moves since you could have written several WriteEndObjects (or even 1) and then the next WriteStartObject would fail because CurrentDepth > 1_000, when it shouldn't.

Reply:
I'm not sure I follow.

If SkipValidation is on (and validation is thus off), no exception is thrown, so by the time the method ends `_currentDepth` should have been adjusted.

My concern is that if `SkipValidation` is off (and validation is thus on), this method has a side effect before the exception is thrown.  The exception should be thrown before the field is modified.

Reply:
> My concern is that if `SkipValidation` is off (and validation is thus on), this method has a side effect before the exception is thrown.

Ah yes, I see what you mean now. You are right, will fix this.